silkworm/rpc/ws/websocket_connection.cpp

```diff
@@ -0,0 +1,124 @@
+   limitations under the License.
+*/
+//
+// Copyright (c) 2003-2020 Christopher M. Kohlhoff (chris at kohlhoff dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#include "websocket_connection.hpp"
+
+#include <exception>
+#include <fstream>
+#include <string_view>
+
+#include <boost/asio/use_awaitable.hpp>
+#include <boost/asio/write.hpp>
+#include <boost/beast/websocket/rfc6455.hpp>
+#include <boost/system/error_code.hpp>
+
+#include <silkworm/infra/common/log.hpp>
+#include <silkworm/rpc/common/util.hpp>
+
+namespace beast = boost::beast;          // from <boost/beast.hpp>
+namespace http = beast::http;            // from <boost/beast/http.hpp>
+namespace websocket = beast::websocket;  // from <boost/beast/websocket.hpp>
+namespace net = boost::asio;             // from <boost/asio.hpp>
+
+namespace silkworm::rpc::http {
+
+WebSocketConnection::WebSocketConnection(boost::beast::websocket::stream<boost::beast::tcp_stream>&& stream,
+                                         commands::RpcApi& api,
+                                         const commands::RpcApiTable& handler_table)
+    : ws_{std::move(stream)},
+      request_handler_{this, api, handler_table} {
+    SILK_DEBUG << "WebSocketConnection::WebSocketConnection ws created:" << &ws_;
+}
+
+WebSocketConnection::~WebSocketConnection() {
+    // socket_.close();
+    SILK_DEBUG << "WebSocketConnection::~WebSocketConnection ws deleted:" << &ws_;
+}
+
+Task<void>
+WebSocketConnection::do_accept(const boost::beast::http::request<boost::beast::http::string_body>& req) {
+    // Set suggested timeout settings for the websocket
+    ws_.set_option(websocket::stream_base::timeout::suggested(beast::role_type::server));
+
+    // Accept the websocket handshake
+    co_await ws_.async_accept(req, boost::asio::use_awaitable);
+}
+
+Task<void> WebSocketConnection::read_loop() {
+    try {
+        while (true) {
+            co_await do_read();
+        }
+    } catch (const boost::system::system_error& se) {
+        if (se.code() == boost::beast::http::error::end_of_stream || se.code() == boost::asio::error::broken_pipe ||
+            se.code() == boost::asio::error::connection_reset || se.code() == boost::beast::websocket::error::closed) {
+            SILK_TRACE << "WebSocketConnection::read_loop close from client with code: " << se.code();
+        } else if (se.code() != boost::asio::error::operation_aborted) {
+            SILK_ERROR << "WebSocketConnection::read_loop system_error: " << se.what();
+            throw;
+        } else {
+            SILK_TRACE << "WebSocketConnection::read_loop operation_aborted: " << se.what();
+        }
+    } catch (const std::exception& e) {
+        SILK_ERROR << "WebSocketConnection::read_loop exception: " << e.what();
+        throw;
+    }
+}
+
+Task<void>
+WebSocketConnection::do_read() {
+    std::string content;
```

Comment:
Probably you can store this buffer as state variable and reserve a maximum space w/o incurring allocation for every call